src/js/directives/highlights.js

```diff
@@ -0,0 +1,138 @@
+var rangy = require('rangy');
+var _ = require('lodash');
+var $ = require('jquery');
+
+module.exports = function (app) {
+
+  // returns all leaf text nodes that are descendants of node or are node
+  var getTextNodes = function (node) {
+    if (!node) return [];
+    if (node.nodeType === Node.TEXT_NODE) return [node];
+    
+    // process childs
+    var nodes = [];
+    $(node).contents().each(function (index, el) {
+      nodes = nodes.concat(getTextNodes(el));
+    });
+    return nodes;
+  };
+
+  app.directive('highlightContainer', function () {
+    return {
+      restrict: 'A',
+      transclude: true,
+      replace: true,
+      scope: {},
+      template: '<div ng-transclude></div>',
+      controller: ['$scope', function ($scope) {
+      }]
+    };
+  });
+  
+  app.directive('highlightRoot', function () {
+    return {
+      restrict: 'A',
+      require: '^^highlightContainer',
```

Comment:
We're requiring the controller of the directive `highlightContainer` here. `^^` means: Locate the required controller by searching the element's parents (see the [docs](https://docs.angularjs.org/api/ng/service/$compile)).